examples/basic_query_extractor/src/main.rs

```diff
@@ -0,0 +1,149 @@
+//! A basic example application for working with the Gotham Query String Extractor 
+
+extern crate futures;
+extern crate gotham;
+extern crate hyper;
+extern crate mime;
+extern crate serde;
+#[macro_use]
+extern crate serde_derive;
+extern crate serde_json;
+#[macro_use]
+extern crate gotham_derive;
+
+use hyper::{Response, StatusCode};
+
+use gotham::http::response::create_response;
+use gotham::router::Router;
+use gotham::router::builder::{build_simple_router, DefineSingleRoute, DrawRoutes};
+use gotham::state::State;
+
+/// `QueryParam` struct
+///
+/// It contains only a `name` field for the sake of simplicity
+/// It permits to use it as query parameter like:
+/// GET /widgets?name=t-shirt
+#[derive(Deserialize, StateData, StaticResponseExtender)]
+struct QueryParam {
+    name: String,
+}
+
+/// `Product` struct
+///
+/// It represents products that can be queried
+#[derive(Serialize)]
+struct Product {
+    name: String,
+    price: f32,
+}
+
+// Generate a `Vec<Product>`
+fn generate_products() -> Vec<Product> {
+    let mut products = Vec::with_capacity(3);
+    products.push(Product {
+        name: "t-shirt".to_string(),
+        price: 15.5,
+    });
+    products.push(Product {
+        name: "sticker".to_string(),
+        price: 1.5,
+    });
+    products.push(Product {
+        name: "mug".to_string(),
+        price: 10.0,
+    });
+    products
+}
+
+/// Function to handle the `GET` requests coming to `/widgets/:name`
+fn get_product_handler(mut state: State) -> (State, Response) {
+    // Extract `name` from `state`
+    let name = state.take::<QueryParam>().name;
```

Comment:
I feel like highlighting the [FromState](https://github.com/gotham-rs/gotham/blob/935ab297113918e050064eab1addcadc619e6ed9/gotham/src/state/from_state.rs) trait in our examples is a good idea, thus this would become:

```
let query_param = QueryParam::borrow_from(&state); 
```